component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -161,58 +160,191 @@ protected void initiateAuthenticationRequest(HttpServletRequest request,
                 }
-                if (StringUtils.isNotEmpty(email)) {
-                    processEmailOTPFlow(request, response, email, username, queryParams, context);
-                }
+
             } else {
-                if (log.isDebugEnabled()) {
-                    log.debug("Process with the first step (basic) authenticator only");
+                // This block need to be revised and recommended to be removed
+                FederatedAuthenticatorUtil.setUsernameFromFirstStep(context);
+                String username = String.valueOf(context.getProperty(EmailOTPAuthenticatorConstants.USER_NAME));
+                authenticatedUser = (AuthenticatedUser) context.getProperty
+                        (EmailOTPAuthenticatorConstants.AUTHENTICATED_USER);
+                // find the authenticated user.
+                if (authenticatedUser == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cannot find the authenticated user, the username : " + username + " may be null");
+                    }
+                    throw new AuthenticationFailedException
+                            ("Authentication failed!. Cannot find the authenticated user, the username : "
+                                    + username + " may be null");
+                }
+
+                boolean isUserExistence = FederatedAuthenticatorUtil.isUserExistInUserStore(username);
+
+                if (isEmailOTPMandatory) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Process the EmailOTP mandatory flow ");
+                    }
+                    processEmailOTPMandatory(context, request, response, isUserExistence, username, queryParams,
+                            emailOTPParameters, sendOtpToFederatedEmail);
+
+                } else if (isUserExistence && !isEmailOTPDisableForUser(username, context,
+                        emailOTPParameters)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Process the EmailOTP optional flow, but user enable emailOTP as second step ");
+                    }
+
+                    email = getEmailValueForUsername(username, context);
+                    if (StringUtils.isEmpty(email)) {
+                        if (request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS) == null) {
+                            redirectToEmailAddressReqPage(response, context, emailOTPParameters, queryParams, username);
+                        } else {
+                            updateEmailAddressForUsername(context, request, username);
+                            email = getEmailValueForUsername(username, context);
+                        }
+                    }
+
+                    if (StringUtils.isNotEmpty(email)) {
+                        processEmailOTPFlow(request, response, email, username, queryParams, context);
+                    }
```

Comment:
Code duplicated. Line 228:243 and 327:339. Shall we extract this to a private method?

Reply:
I left it as we planned to revise and remove this block of code. will get it done.

Reply:
done